### ISSUE TITLE
LUIS Bot Integrated with Slack

### ISSUE THREAD

**Issue by anhnh305** [microsoft/BotFramework-Composer#4480]
I've created some bots with Bot Composer and test it out fine on both Emulator and Web Chat. But when I connect it to my Slack channel, these two answer just came out of no where. The first one "Value ..." is the bot which I have those custom action to get data from DW and show as an answer (report/chart). 

The second one is when I connect with another bot just to test it out if LUIS is working or not (Simple bot that have some questions with Default Recognizer in Composer). I'm running out of idea, does anyone know what is the problem because I'm pretty sure it's just about the connect to slack thingy.

P/s: Both Bots work fine with if they're all set to regular expression.

![Screenshot 2020-10-22 235548](https://user-images.githubusercontent.com/44403719/96905073-29a7b600-14c2-11eb-8ed1-931c718946ac.png)

**Comment by v-kydela** [comment]
@anhnh305 - This looks like it could be related to the multi-language recognizer problem: https://github.com/microsoft/BotFramework-Composer/issues/4180

What version of Composer are you using?

**Comment by anhnh305** [comment]
@v-kydela I'm currently using the 1.1.1 version
![image](https://user-images.githubusercontent.com/44403719/96943105-49120380-1501-11eb-8f9f-78c6e0be6e42.png)

**Comment by v-kydela** [comment]
@anhnh305 - So you are using the desktop version of Composer and not the browser-based version, right? This is important to know, because if you build Composer yourself from source code then you'll have all the latest fixes. If you use the desktop version then you'll only have the fixes made before the last installer was made.

The last time I addressed the multi-language recognizer problem I provided three options:

1. Wait for an update to Composer
2. Use a custom runtime with updated Bot Builder packages
3. Bypass the multi-language recognizer by manually editing your .dialog file

**Comment by anhnh305** [comment]
@v-kydela Can you please show me how to manually editing my .dialog file to bypass the multi-language recognizer or some documentation to follow. I'm not very clear about that, thank you.

**Comment by v-kydela** [comment]
@anhnh305 - There's no documentation for how to hack your Composer bot to get around a problem that's already been fixed, of course. And I can't show you how to edit your file when I haven't even seen your file. Perhaps you'd like to provide the files in question.

**Comment by anhnh305** [comment]
@v-kydela I'm sorry if I confusing you or make you feel disrespectful but this is my .dialog file's value. The reason why I ask that question is because I don't seem to find any part in the .dialog file that related to the multi-language or so, and I can't find any document to find out more about it. Maybe I should add a block and state that this bot is only in en-us? Please give me some advices, thank you

.dialog file:
{
  "$kind": "Microsoft.AdaptiveDialog",
  "$designer": {
    "$designer": {
      "name": "ParacelVN",
      "description": "A bot that can help with retail problem",
      "id": "zjI5Zx"
    }
  },
  "defaultResultProperty": "dialog.result",
  "triggers": [
    {
      "$kind": "Microsoft.OnConversationUpdateActivity",
      "$designer": {
        "id": "376720"
      },
      "actions": [
        {
          "$kind": "Microsoft.SendActivity",
          "$designer": {
            "id": "NneL7d"
          },
          "activity": "${SendActivity_NneL7d()}"
        },
        {
          "$kind": "Microsoft.Foreach",
          "$designer": {
            "id": "518944",
            "name": "Loop: for each item"
          },
          "itemsProperty": "turn.Activity.membersAdded",
          "actions": [
            {
              "$kind": "Microsoft.IfCondition",
              "$designer": {
                "id": "641773",
                "name": "Branch: if/else"
              },
              "condition": "string(dialog.foreach.value.id) != string(turn.Activity.Recipient.id)"
            }
          ],
          "value": "dialog.foreach.value",
          "index": "dialog.foreach.index"
        }
      ]
    },
    {
      "$kind": "Microsoft.OnIntent",
      "$designer": {
        "id": "haXioK"
      },
      "intent": "Inventory",
      "actions": [
        {
          "$kind": "Microsoft.SendActivity",
          "$designer": {
            "id": "ZCt0CV"
          },
          "activity": "${SendActivity_ZCt0CV()}"
        }
      ]
    },
    {
      "$kind": "Microsoft.OnIntent",
      "$designer": {
        "id": "9wryif"
      },
      "intent": "Causing",
      "actions": [
        {
          "$kind": "Microsoft.SendActivity",
          "$designer": {
            "id": "Ay5ZL6"
          },
          "activity": "${SendActivity_Ay5ZL6()}"
        },
        {
          "$kind": "Microsoft.SendActivity",
          "$designer": {
            "id": "WRR6JA"
          },
          "activity": "${SendActivity_WRR6JA()}"
        }
      ]
    },
    {
      "$kind": "Microsoft.OnIntent",
      "$designer": {
        "id": "4vMjjW"
      },
      "intent": "Solution",
      "actions": [
        {
          "$kind": "Microsoft.SendActivity",
          "$designer": {
            "id": "X2QtW2"
          },
          "activity": "${SendActivity_X2QtW2()}"
        }
      ]
    },
    {
      "$kind": "Microsoft.OnIntent",
      "$designer": {
        "id": "Jvr4nA"
      },
      "intent": "Request",
      "actions": [
        {
          "$kind": "Microsoft.SendActivity",
          "$designer": {
            "id": "Alo7Di"
          },
          "activity": "${SendActivity_Alo7Di()}"
        },
        {
          "$kind": "Microsoft.SendActivity",
          "$designer": {
            "id": "UgurNo"
          },
          "activity": "${SendActivity_UgurNo()}"
        }
      ]
    },
    {
      "$kind": "Microsoft.OnIntent",
      "$designer": {
        "id": "dNy3Dz"
      },
      "intent": "Timestamp",
      "actions": [
        {
          "$kind": "Microsoft.TextInput",
          "$designer": {
            "id": "qoRf3a"
          },
          "disabled": false,
          "maxTurnCount": 3,
          "alwaysPrompt": false,
          "allowInterruptions": false,
          "prompt": "${TextInput_Prompt_qoRf3a()}",
          "property": "conversation.product"
        },
        {
          "$kind": "Microsoft.DateTimeInput",
          "$designer": {
            "id": "HJmVfl"
          },
          "disabled": false,
          "maxTurnCount": 3,
          "alwaysPrompt": false,
          "allowInterruptions": false,
          "prompt": "${DateTimeInput_Prompt_HJmVfl()}",
          "unrecognizedPrompt": "${DateTimeInput_UnrecognizedPrompt_HJmVfl()}",
          "defaultValueResponse": "${DateTimeInput_DefaultValueResponse_HJmVfl()}",
          "property": "user.date"
        },
        {
          "$kind": "Microsoft.DateTimeInput",
          "$designer": {
            "id": "Iun6rU"
          },
          "disabled": false,
          "maxTurnCount": 3,
          "alwaysPrompt": false,
          "allowInterruptions": false,
          "prompt": "${DateTimeInput_Prompt_Iun6rU()}",
          "unrecognizedPrompt": "${DateTimeInput_UnrecognizedPrompt_Iun6rU()}",
          "defaultValueResponse": "${DateTimeInput_DefaultValueResponse_Iun6rU()}",
          "property": "user.date2"
        },
        {
          "$kind": "Microsoft.SendActivity",
          "$designer": {
            "id": "BlSLSO"
          },
          "activity": "${SendActivity_BlSLSO()}"
        },
        {
          "$kind": "MultiplyDialog",
          "$designer": {
            "id": "EdVyU6"
          },
          "arg1": 30,
          "arg2": 5,
          "resultProperty": "dialog.result"
        },
        {
          "$kind": "Microsoft.SendActivity",
          "$designer": {
            "id": "raS7nY"
          },
          "activity": "${SendActivity_raS7nY()}"
        }
      ]
    },
    {
      "$kind": "Microsoft.OnIntent",
      "$designer": {
        "id": "0RcKcf",
        "name": "NetSaleDaily"
      },
      "intent": "NetSaleDaily",
      "actions": [
        {
          "$kind": "Microsoft.SendActivity",
          "$designer": {
            "id": "7OYFDI"
          },
          "activity": "${SendActivity_7OYFDI()}"
        }
      ]
    }
  ],
  "$schema": "https://raw.githubusercontent.com/microsoft/BotFramework-Composer/stable/Composer/packages/server/schemas/sdk.schema",
  "generator": "paracelvn.lg",
  "recognizer": "paracelvn.lu.qna",
  "autoEndDialog": true,
  "id": "paracelvn"
}

**Comment by v-kydela** [comment]
@anhnh305 - Thank you for the .dialog file. The relevant piece of information in the JSON you've posted is the recognizer property, which is `"paracelvn.lu.qna"`. You may see a file called paracelvn.lu.qna.dialog in a folder called generated, but if it's not there then don't worry because Composer generates it automatically. If it is there then you can find the definition of your recognizer inside it, which will likely be a cross-trained recognizer set. That might point you to another file that contains a definition for a multi-language recognizer, which in turn should finally point you to the language-specific LUIS recognizer.

If you have a language-understanding folder which contains an en-us folder which contains a file called paracelvn.en-us.lu, you can just replace the recognizer in your .dialog file with paracelvn.en-us.

Does this resolve your issue?

**Comment by anhnh305** [comment]
Hi @v-kydela. After changing the CrossTrainedRecognizerSet like you said's value to paracelvn.en-us. My LUIS test bot still got the same error, which is:
![image](https://user-images.githubusercontent.com/44403719/97731083-474ece00-1b07-11eb-8fd5-905713abc92a.png)

Further than that, my StaggingBot which involve with data couldn't even publish to Azure to test when I change the value in my .dialog inside "generated" folder. It's struggled at choosing which .lu file to use I assume:
![image](https://user-images.githubusercontent.com/44403719/97731429-add3ec00-1b07-11eb-83e6-68ec8ff06196.png)

Any suggestion?

KRs

**Comment by anhnh305** [comment]
@v-kydela I was misunderstood you for a sec, the first part of your answer is to find the multi-language file, I thought that's the .dialog file to fix the value inside (which is a CrossTrained...'s value). So when I finally realize what am I doing wrong, I fix the recognizer value in the .dialog file right under the bot's folder (bot-name.dialog). Then a bunch of files with bot's name prefix in the generated folder is gone (only left with bot-name.lu.qna.dialog & chatlog.lu.qna.dialog file). My bot couldn't even test in the emulator as well as in web chat (when I published it into azure later).

TLDR: I found the multi-language file. I tried to change the value of the recognizer in bot's .dialog file and then it's couldn't test, stuck in "POST500directline/conversations/<conversationId>/activities": 

![image](https://user-images.githubusercontent.com/44403719/97785213-36b95900-1bd6-11eb-9db0-46329e7ebcd2.png)

I'm very sorry if my bad english is disturbing you but please help me find the solution. Thanks

**Comment by v-kydela** [comment]
@anhnh305 - I don't have enough information about your Composer project to troubleshoot this for you. If you can't get option 3 to work then you'll have to go with option 1 or 2

**Comment by anhnh305** [comment]
@v-kydela Thanks, I got it, but in case I got the latest version of bot builder as well as Bot Composer, didn't it mean I have to wait for either update of Composer nor Bot Builder?

**Comment by v-kydela** [comment]
@anhnh305 - The latest version of Composer does not use the latest version of Bot Builder. You will need to wait until Composer is updated to use a version of Bot Builder with the fix to the multi-language recognizer problem. Does this resolve your issue?

**Comment by anhnh305** [comment]
@v-kydela Well currently I'm still waiting for the update of Composer to see if it solve my issue or not. But I'm still not very clear about the third method you gave me. Because:
+ I located the MultiLanguageRecognizer file which is my botname.lu.dialog inside the generated folder and it's show like this:
{
    "$kind": "Microsoft.MultiLanguageRecognizer",
    "id": "LUIS_assisty",
    "recognizers": {
        "en-us": "luisbottest.en-us.lu",
        "": "luisbottest.en-us.lu"
    }
}
+ I then changing the botname.dialog right under the bot's folder's recognizer value like you said to botname.en-us. And whenever I do that, the bot will stop working in the emulator as well as on the Azure Portal Webchat (therefore no such thing as working in Slack). The Bot Emulator will show this
![image](https://user-images.githubusercontent.com/44403719/98334858-6c2ad000-2036-11eb-8980-c7fc9107277c.png)

Funny thing is, if I change back the recognizer value in the botname.dialog as default which is "botname.lu.qna". Everything will comeback to normal, work on Emulator, Azure but still not in Slack.
![image](https://user-images.githubusercontent.com/44403719/98335273-3fc38380-2037-11eb-9020-95d3514da3b3.png)

So basically, the only way you show me that I can do to fix the issue (the third method) is causing the bot to not working. If anything I do wrong please tell me, I've been stuck with this for weeks now. Really appreciate your helps the past few weeks.

**Comment by v-kydela** [comment]
@anhnh305 - I've just confirmed that botname.en-us will not work and you need to use botname.en-us.lu, the same format that you saw in the MultiLanguageRecognizer. Does this resolve your issue?

The multi-language recognizer fix is already in the Composer source code, so you should be fine if you use the latest Composer code by building it yourself according to the instructions. The Electron build is using code that's a few months old, and if you want to make sure you're using the latest code then you should always build directly from source.

@cwhitten - When is the next Electron release of Composer?

**Comment by anhnh305** [comment]
@v-kydela I've fixed the issue by changing the recognizer value to botname.en-us.lu. Thank you so much for the past few weeks has stuck with my problem.
